packages/rpc-transport/src/apis/api-types.ts

```diff
@@ -0,0 +1,11 @@
+export type RpcApiConfig = Readonly<{
+    parametersTransformer?: <T>(params: T, methodName: string) => unknown[];
+    responseTransformer?: <T>(response: unknown, methodName: string) => T;
+}>;
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type RpcMethod = (...args: any) => any;
+
+export interface IRpcApiMethods {
+    [methodName: string]: RpcMethod;
```

Comment:
I don't think this actually limits the type as you wish. [Playground link](https://www.typescriptlang.org/play?#code/MYewdgzgLgBARgJwIZgCYwLwwMoE8C2cIANgBQCUA3AFDUCWYUApggGZLBMwCSASgA7AAskygALEOgDe1GHJgBtRClQBdAFwwwTAG4tqAX1oNmbDlwBiIECPGSYTAB7M0EHgOGiJ02fNbWKTR0QOlRDY0YWdk4YACEkBFtvB2cmV3dBJPsZeXgEwJhg0PDqUEhYaABXVlZMLUriYhgkN0qwAGswEAB3MGa3eMSvewAfGCsbYdQaKpqYCCQoOghWOiY3PkypmloAen2Dw73dhwhiEwBaVGWkOGImC+1nC-PtGAABKFx+deAEOn4UAu61eUF2XWBjn452AdCBKFw1C+PxgHiy6CwpAAdDiEgBzCCaBHkTAAPmaYFwOxMUXMGU8dlQAHlxCwYDl5Ap8FNNNB-mA8RpUVtGSUaWYYvEAF7olIuVAbNFTFliNkcuRwJBSgpFMJGeiRCVceJQWXqxTcxm8qD8wWaChkwohPW0UrgaAwcQMAl1MANJotGBtTo9PqB6WysYm9E0L0CtwLJYrNaKkXeFUsSgwfYwaz8CDUIA).

What you want is a ‘marker’ interface – an empty interface that does literally nothing and specifies literally nothing other than to make a bunch of implementers ‘participate’ in it.

Reply:
I hated this approach but I did try this initially 💀

Reply:
Take a look at my playground. I added a solution there.

In particular, try removing `extends IRpcMethod` from `interface BarMethod extends IRpcMethod` and you'll see it raises a type error.

Reply:
Right I can see how you did it, and the interface I wrote works the same way. All of the methods don't _need_ to extend `IRpcApiMethods`, but it felt weird to just make the first one extend it and then none of the rest. I figured people would be like (..?).

Specifically, I can mimic your playground by _just_ making, for example, `GetAccountInfoApi` extend `IRpcApiMethods` and none of the rest. That results in a valid generic type arg `T` for `createJsonRpcApi<T>(..)`.

Reply:
I also don't want to just add some random interface that does nothing, because that results in a property on the API object, which does nothing, and that's not kewl.

Reply:
> …it felt weird to just make the first one extend it and then none of the rest.

I'm confused. My playground is a bit incoherent, maybe. Let me explain.

There are two parts to it, separated by `//////////`. The bottom part is the bad example, the top the good.

The bad example shows that you can _forget_ to add the marker interface to the second typedef (`BatMethod`) and the thing will still typecheck. That's the thing we want to avoid. All of the typedefs _should_ extend the marker interface.

The good example shows that _if_ you remove the marker interface from even one of the typedefs, a type error is raised. This is what we want – to enforce that anything that participates in that union extends the marker interface `IRpcMethod`.

I'm not just making stuff up; marker interfaces are [a thing](https://en.wikipedia.org/wiki/Marker_interface_pattern).

> …that results in a property on the API object, which does nothing, and that's not kewl.

It doesn't!

![image.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/lE1zD7Jpg6mWNv3djNKZ/45bcfe99-3500-4c3c-b173-00d35b523372.png)

Reply:
> In particular, try removing `extends IRpcMethod` from `interface BarMethod extends IRpcMethod` and you'll see it raises a type error.

Hey, unless I'm missing something, this isn't the actual behavior. [Playground link](https://www.typescriptlang.org/play?#code/MYewdgzgLgBARgJwIZgCYwLwwMoE8C2cIANgBQCUA3AFDUCWYUApggGZLBMwCSASgA7AAskygALEOgDe1GHJgBtRClQBdAFwwwTAG4tqAX1oNmbDlwBiIECPGSYTAB7M0EHgOGiJ02fNbWKTR0QOlRDWgB6CJgAAycXVAgYmAQmfBA9MJMWdk4YACEkBFtvGBl5eCLAmGDQ8OpQSFhoAFdWVkwtFuJiGCQ3FrAAazAQAHcwPrdC4q97ADIYKxs51BpW9pgIJCg6CFY6Jjc+QRLJShgomGt+CCA)

Reply:
Uhhh. OK, there are a few things wrong with my sort of braindead playground.

1. Yeah obviously the intersection of the two satisfies `IRpcMethod` (duh); oops – my bad.
2. It's not actually interesting that the _intersection_ of the interfaces satisfies `IRpcMethod`, it's important that each one individually satisfies it.
3. If you change that `&` to a `|` you'll see the type error.

Anyway, this has gone on long enough for nothing. Let's pretend I never said anything; I don't think this typing is adding any value.